### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by kzuchowska-splunk**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [ ] Testing of all the changes has been performed (for bug fixes / features)
- [ ] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [ ] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [ ] Verify all checks are passing.
- [ ] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [ ] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,84 @@
-# Splunk> Phantom
+[comment]: # "Auto-generated SOAR connector documentation"
+# Discord
 
-Welcome to the open-source repository for Splunk> Phantom's discord App.
+Publisher: discord
```
Comment (ivellios) [inline README.md]: We are the ones who publish the app, while Discord is a vendor
```suggestion
Publisher: Splunk
```

File: README.md [hunk 2]
```diff
@@ -1,9 +1,84 @@
...
-Welcome to the open-source repository for Splunk> Phantom's discord App.
+Publisher: discord  
+Connector Version: 1.0.0  
+Product Vendor: Discord  
+Product Name: Discord  
+Product Version Supported (regex): ".\*"  
+Minimum Product Version: 6.2.2.134  
 
-Please have a look at our [Contributing Guide](https://github.com/Splunk-SOAR-Apps/.github/blob/main/.github/CONTRIBUTING.md) if you are interested in contributing, raising issues, or learning more about open-source Phantom apps.
+Integrate with Discord to post messages and attachments to channels
```
Comment (ivellios) [inline README.md:11]: Do we actually post attachments? I think we rather can analyze them (so the other way around).

File: README.md [hunk 3]
```diff
@@ -1,9 +1,84 @@
...
 
-## Legal and License
+### Configuration Variables
+The below configuration variables are required for this Connector to operate.  These variables are specified when configuring a Discord asset in SOAR.
 
-This Phantom App is licensed under the Apache 2.0 license. Please see our [Contributing Guide](https://github.com/Splunk-SOAR-Apps/.github/blob/main/.github/CONTRIBUTING.md#legal-notice) for further details.
+VARIABLE | REQUIRED | TYPE | DESCRIPTION
+-------- | -------- | ---- | -----------
+**token** |  required  | password | Discord bot token
+**guild_id** |  required  | numeric | Guild ID
```
Comment (ivellios) [inline README.md]: I would provide additional URL to the docs in Discord, like:
- https://support.discord.com/hc/en-us/articles/23187611406999-Guilds-FAQ
- https://support.discord.com/hc/en-us/articles/206346498-Where-can-I-find-my-User-Server-Message-ID#h_01HRSTXPS5FSFA0VWMY2CKGZXA -- how to get the Guild ID 

Also, I would rather put some note like "aka Server ID"

File: discord.json [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        }
+    ],
+    "type": "information",
+    "product_vendor": "Discord",
+    "logo": "discord.svg",
+    "logo_dark": "discord_dark.svg",
+    "product_name": "Discord",
+    "python_version": "3",
+    "product_version_regex": ".*",
+    "publisher": "discord",
```
Comment (ivellios) [inline discord.json]:
```suggestion
    "publisher": "Splunk",
```

File: discord.json [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    "license": "Copyright (c) Discord, 2024",
```
Comment (ivellios) [inline discord.json]:
```suggestion
    "license": "Copyright (c) 2024 Splunk Inc.",
```

File: README.md [hunk 6]
```diff
@@ -1,9 +1,84 @@
...
+Integrate with Discord to post messages and attachments to channels
 
-## Legal and License
+### Configuration Variables
+The below configuration variables are required for this Connector to operate.  These variables are specified when configuring a Discord asset in SOAR.
 
-This Phantom App is licensed under the Apache 2.0 license. Please see our [Contributing Guide](https://github.com/Splunk-SOAR-Apps/.github/blob/main/.github/CONTRIBUTING.md#legal-notice) for further details.
+VARIABLE | REQUIRED | TYPE | DESCRIPTION
+-------- | -------- | ---- | -----------
+**token** |  required  | password | Discord bot token
```
Comment (ivellios) [inline README.md:18]: Some URL to the discord support page? Maybe: https://support-dev.discord.com/hc/en-us/articles/6470840524311-Why-can-t-I-copy-my-bot-s-token

File: discord.json [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            "read_only": true,
+            "parameters": {},
+            "output": [],
+            "versions": "EQ(*)"
+        },
+        {
+            "action": "list channels",
+            "identifier": "list_channels",
+            "description": "List text channels of a guild",
+            "verbose": "The output of this action is a list of all text channels of a guild. The channels will be listed with their corresponding channel IDs.",
```
Comment (ivellios) [inline discord.json]:
```suggestion
            "verbose": "The output of this action is a list of all text channels for the guild. The channels will be listed with their corresponding channel IDs.",
```

File: discord.json [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                "title": "List Channels",
+                "type": "table",
+                "height": 5
+            },
+            "versions": "EQ(*)"
+        },
+        {
+            "action": "send message",
+            "identifier": "send_message",
+            "description": "Send a message to Discord",
```
Comment (ivellios) [inline discord.json]:
```suggestion
            "description": "Send a message to the Discord channel",
```

File: discord_connector.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+# Phantom sample App Connector python file
+# -----------------------------------------
+
+# Phantom App imports
+import phantom.app as phantom
+from phantom.action_result import ActionResult
+from phantom.base_connector import BaseConnector
+
+# Usage of the consts file is recommended
+# from discord2_consts import *
```
Comment (ivellios) [inline discord_connector.py]: leftovers

File: discord_connector.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    def __new__(cls, val1, val2=None):
+        return tuple.__new__(RetVal, (val1, val2))
+
+
+class Discord2Connector(BaseConnector):
+
+    def __init__(self):
+
+        # Call the BaseConnectors init first
+        super(Discord2Connector, self).__init__()
```
Comment (ivellios) [inline discord_connector.py]:
```suggestion
        super().__init__()
```
At this point there's no need to stick to the pre py3 syntax ;)

File: discord_connector.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+                    phantom.APP_ERROR, "Error Connecting to server. Details: {0}".format(str(e))
+                ), resp_json
+            )
+
+        return self._process_response(r, action_result)
+
+    def _handle_test_connectivity(self, param):
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        self._loop.run_until_complete(self._client.login(self._token))
```
Comment (ivellios) [inline discord_connector.py]: Maybe we could put `_loop` also in the constructor, so it is clear, that this is used?

File: discord_connector.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+
+        if self._client.status != discord.Status.online:
+            self.save_progress("Test Connectivity Failed.")
```
Comment (ivellios) [inline discord_connector.py]: Maybe we can also add some `set_status` on the result with some details on the connectivity issue (the actual status code, maybe some message if available)?

File: discord_connector.py [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+            return action_result.get_status()
+
+        self.save_progress("Test Connectivity Passed")
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    async def _handle_get_guild(self):
```
Comment (ivellios) [inline discord_connector.py]: is this actually the action handler (looking at the name). Perhaps this should be more like `fetch_guild` or `load_guild` or even `preload_guild` - to mark that it is being loaded into `self._guild` and not returned or handled in any way.

File: discord_connector.py [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+        channels = self._loop.run_until_complete(self._guild.fetch_channels())
+        num_channels = 0
+
+        for channel in channels:
+            if type(channel) == discord.TextChannel:
+                num_channels += 1
+                add_chan = {}
+                add_chan['name'] = channel.name
+                add_chan['id'] = channel.id
+                action_result.add_data(add_chan)
```
Comment (ivellios) [inline discord_connector.py]: Can be simplified into:
```suggestion
                action_result.add_data({
                    "name": channel.name,
                    "id": channel.id
                })
```

File: discord_connector.py [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        self.save_progress(str(type(self._guild)))
+
+        channels = self._loop.run_until_complete(self._guild.fetch_channels())
+        num_channels = 0
+
+        for channel in channels:
+            if type(channel) == discord.TextChannel:
+                num_channels += 1
```
Comment (ivellios) [inline discord_connector.py]: Is `channels` a genarator? Or we can simply replace this with len(channels) below in :198 ?

File: discord_connector.py [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+    def _handle_send_message(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        destination = param['destination']
+        message = param['message']
+
+        channel = self._loop.run_until_complete(self._guild.fetch_channel(destination))
+        self._loop.run_until_complete(channel.send(message))
```
Comment (ivellios) [inline discord_connector.py]: can these two run within one `run_until_complete`?

File: discord_connector.py [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+
+        self._loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self._loop)
+
+        try:
+            self._loop.run_until_complete(self._handle_get_guild())
+        except discord.LoginFailure:
+            self.save_progress("Login Failure")
+        except discord.NotFound:
+            self.save_progress("Guild not found")
```
Comment (ivellios) [inline discord_connector.py]: Shouldn't this method also return `phantom.APP_ERROR` on any issue?

**Review by ivellios — CHANGES_REQUESTED** [review]
Overall, nice init. Some things to consider, some to change :) 

I left some comments on README, but AFAIR it's being automatically generated?

File: discord_connector.py (hunk 15 above)
Comment (kzuchowska-splunk) [reply discord_connector.py]: channels is a list of all channels including text voice etc, so we cannot use len(channels)

File: discord_connector.py (hunk 16 above)
Comment (kzuchowska-splunk) [reply discord_connector.py]: now we use run_in_loop function

File: discord_connector.py [hunk 18]
```diff
@@ -0,0 +1,566 @@
...
+    def initialize(self):
+        # Load the state in initialize, use it to store data
+        self._state = self.load_state()
+        # get the asset config
+        config = self.get_config()
+
+        self._base_url = "https://discord.com/api/v10"
+        self._token = config['token']
+        self._guild_id = config['guild_id']
+        self._headers = {"Authorization": "Bot " + self._token}
```
Comment (mposluszny-splunk) [inline discord_connector.py]: I can't see _headers used anywhere

File: discord_connector.py [hunk 19]
```diff
@@ -0,0 +1,566 @@
...
+
+        intents = discord.Intents.default()
+        intents.presences = True
+        intents.members = True
+        intents.message_content = True
```
Comment (mposluszny-splunk) [inline discord_connector.py]: isn't it the same as `discord.Intents.all()`?

File: discord_connector.py [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+
+
+    def _handle_list_channels(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        status, channels = self.run_in_loop(self._guild.fetch_channels(), action_result,
+                                            message="Cannot fetch channel from Discord.")
+
+        num_channels = 0
```
Comment (mposluszny-splunk) [inline discord_connector.py]: I'd change that to `text_channels_count` it's a bit more descriptive, as we don't want to gather other type of channels

File: discord_connector.py [hunk 21]
```diff
@@ -0,0 +1,566 @@
...
+        self.save_progress("Test Connectivity Passed")
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def _handle_fetch_message(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        channel_id = param['channel_id']
+        message_id = param['message_id']
```
Comment (mposluszny-splunk) [inline discord_connector.py]: do we need to unpack dict values to variables if we are not using them more than once?

File: discord_connector.py [hunk 22]
```diff
@@ -0,0 +1,566 @@
...
+            "name": user.name,
+            "created_at": str(user.created_at),
+            "system": user.system,
+            "public_flags": user.public_flags.all()
+        })
+
+
+        return status
+
+    def run_in_loop(self, coroutine, action_result, message = ""):
```
Comment (mposluszny-splunk) [inline discord_connector.py]: Maybe we could rename `message` to `error_message` to make it  more inline with variable purpose

File: discord_connector.py [hunk 23]
```diff
@@ -0,0 +1,566 @@
...
+            "message data": {
+                "created at": str(message.created_at),
+                "edited at": str(message.edited_at) if message.edited_at is not None else "message was not edited",
+            },
+            "author data": {
+                "author id": message.author.id,
+                "author name": message.author.name,
+            },
+            "jump url": message.jump_url,
+            "flags": list(filter(lambda flag: getattr(message.flags, flag), MESSAGE_FLAGS)) or "no flags",
```
Comment (mposluszny-splunk) [inline discord_connector.py]: Do we need to have this list - MESSAGE_FLAGS? Only thing here that we don't want to show is `value` property (https://discordpy.readthedocs.io/en/latest/api.html#messageflags)

Also, iterator mentioned in the docs should give you a pair of flags, so you don't need to use kinda hackhy `getattr`

**Review by mposluszny-splunk — CHANGES_REQUESTED** [review]

File: discord_connector.py (hunk 20 above)
Comment (kzuchowska-splunk) [reply discord_connector.py]: Instead I used 
```summary['num_channels'] = sum(isinstance(channel, discord.TextChannel) for channel in channels)```
so we don't need to use any variable name

File: discord_connector.py (hunk 21 above)
Comment (kczernik-splunk) [reply discord_connector.py]: It is standardized way of doing things by app wizzard, we could change that but it would mess with the given convention.

Comment (mposluszny-splunk) [reply discord_connector.py]: Good point, let's leave it for refactoring later

File: discord_connector.py (hunk 23 above)
Comment (kczernik-splunk) [reply discord_connector.py]: It has been changed so now it does not use const and getattr

File: discord_connector.py [hunk 24]
```diff
@@ -0,0 +1,436 @@
...
+
+    def _handle_list_channels(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        status, channels = self.run_in_loop(self._guild.fetch_channels(), action_result,
+                                            error_message=DISCORD_ERROR_FETCHING_CHANNEL)
+
+        for channel in channels:
+            if isinstance(channel, discord.TextChannel):
```
Comment (mposluszny-splunk) [inline discord_connector.py]: maybe there is some way in Discord SDK to check channel type without checking objects class? It would be a bit more cleaner

Comment (mposluszny-splunk) [reply discord_connector.py]: also maybe we could do filtering once by keeping text_channels in separate variable? Whats your mind on it?

File: discord_connector.py [hunk 25]
```diff
@@ -0,0 +1,436 @@
...
+            "message data": {
+                "created at": str(message.created_at),
+                "edited at": str(message.edited_at) if message.edited_at is not None else "message was not edited",
+            },
+            "author data": {
+                "author id": message.author.id,
+                "author name": message.author.name,
+            },
+            "jump url": message.jump_url,
+            "flags": [flag[:1] for flag in list(filter((lambda flag: flag[1]), message.flags))] or "no flags",
```
Comment (mposluszny-splunk) [inline discord_connector.py]: why :1 ? Isn't it just 0?

Comment (mposluszny-splunk) [reply discord_connector.py]: Also, why do you need to change filter to list and why are there two pairs brackets (( ))?

**Review by mposluszny-splunk — CHANGES_REQUESTED** [review]

File: discord_connector.py (hunk 25 above)
Comment (kczernik-splunk) [reply discord_connector.py]: :1 is there becouse the result of list(filter(lamba is list of tuples from with we need only the first value

Comment (kczernik-splunk) [reply discord_connector.py]: I just figured out what you ment by :1, I am going to change that

File: discord_connector.py (hunk 24 above)
Comment (kzuchowska-splunk) [reply discord_connector.py]: I don't see any better way to check channel type in discord sdk

Comment (kzuchowska-splunk) [reply discord_connector.py]: I can use
```text_channels = [channel for channel in channels if isinstance(channel, discord.TextChannel)]```
to keep text channels in seperate variable and use it to add action results

Comment (mposluszny-splunk) [reply discord_connector.py]: What about https://discordpy.readthedocs.io/en/stable/api.html?highlight=textchannel#discord.TextChannel.type ?

File: discord_artifact.py [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+    URL: str = ""
+    Type: str = ""
+    Description: str = ""
+
+
+@dataclass
+class Artifact:
+    container_id: str
+    name: str
+    cef: dict = field(default_factory=dict)
```
Comment (ivellios) [inline discord_artifact.py:29]: shouldn't this be `Cef` by default, instead of dict?

File: discord_connector.py [hunk 27]
```diff
@@ -0,0 +1,436 @@
...
+
+    def create_attachment_artifact(self, attachment, container_id):
+        artifact = Artifact(
+            container_id=container_id,
+            name=f"attachment: {attachment.filename}",
+            cef={"URL": attachment.url, "Description": attachment.description, "Type": attachment.content_type}
+        )
+        return self.save_artifact_to_soar(dataclasses.asdict(artifact))
+
+    def save_artifact_to_soar(self, artifact):
```
Comment (ivellios) [inline discord_connector.py]: Wouldn't it be better to accept artifact as instance of `Artifact` instead of dict? This is pretty confusing. See :147

**Review by ivellios — CHANGES_REQUESTED** [review]
Please, consider adding typehints for the methods. The types are now a bit messed up.

**Review by ivellios — DISMISSED** [review]

**Review by ivellios — DISMISSED** [review]

**Review by ivellios — APPROVED** [review]

**Review by mposluszny-splunk — APPROVED** [review]

**Pull request merged**